File: backend/app/schemas/collaboration.py
"""Workflow collaboration API contracts."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

CLIENT_ID_PATTERN = r"^[A-Za-z0-9._:-]+$"


class CollaborationClient(BaseModel):
    client_id: str = Field(min_length=8, max_length=128, pattern=CLIENT_ID_PATTERN)


class CollaborationHeartbeat(CollaborationClient):
    lease_id: str | None = Field(default=None, min_length=32, max_length=128)


class CollaborationLockRequest(CollaborationClient):
    takeover: bool = False
    lease_id: str | None = Field(default=None, min_length=32, max_length=128)


class CollaborationLockRelease(CollaborationClient):
    lease_id: str = Field(min_length=32, max_length=128)


class CollaborationPresenceOut(BaseModel):
    client_id: str
    subject: str
    last_seen_at: datetime
    expires_at: datetime
    is_self: bool


class CollaborationLockOut(BaseModel):
    client_id: str
    subject: str
    acquired_at: datetime
    expires_at: datetime
    owned_by_self: bool


class CollaborationSnapshotOut(BaseModel):
    workflow_id: str
    revision: int
    can_edit: bool
    participants: list[CollaborationPresenceOut] = Field(default_factory=list)
    lock: CollaborationLockOut | None = None
    lease_id: str | None = None
    backend: str = "memory"
    read_only: bool = False

File: frontend/src/api/endpoints/collaboration.ts
/** Workflow presence and soft-lock API endpoints. */

import { apiGet, apiSend } from "@/api/client";
import { withQuery } from "@/api/pagination";

export interface CollaborationPresenceDTO {
  client_id: string;
  subject: string;
  last_seen_at: string;
  expires_at: string;
  is_self: boolean;
}

export interface CollaborationLockDTO {
  client_id: string;
  subject: string;
  acquired_at: string;
  expires_at: string;
  owned_by_self: boolean;
}

export interface CollaborationSnapshotDTO {
  workflow_id: string;
  revision: number;
  can_edit: boolean;
  participants: CollaborationPresenceDTO[];
  lock: CollaborationLockDTO | null;
  lease_id: string | null;
  /** Collaboration coordination backend: memory | redis | unavailable. */
  backend?: string;
  /** True when shared Redis is required but currently unusable. */
  read_only?: boolean;
}

function basePath(workflowId: string): string {
  return `/api/workflows/${workflowId}/collaboration`;
}

export function getWorkflowCollaboration(
  workflowId: string,
  clientId: string,
): Promise<CollaborationSnapshotDTO> {
  return apiGet<CollaborationSnapshotDTO>(
    withQuery(basePath(workflowId), { client_id: clientId }),
  );
}

export function heartbeatWorkflowCollaboration(
  workflowId: string,
  clientId: string,
  leaseId?: string | null,
): Promise<CollaborationSnapshotDTO> {
  return apiSend<CollaborationSnapshotDTO>(
    `${basePath(workflowId)}/heartbeat`,
    "POST",
    { client_id: clientId, ...(leaseId ? { lease_id: leaseId } : {}) },
  );
}

export function acquireWorkflowLock(
  workflowId: string,
  clientId: string,
  options: { takeover?: boolean; leaseId?: string | null } = {},
): Promise<CollaborationSnapshotDTO> {
  return apiSend<CollaborationSnapshotDTO>(
    `${basePath(workflowId)}/lock`,
    "PUT",
    {
      client_id: clientId,
      takeover: options.takeover ?? false,
      ...(options.leaseId ? { lease_id: options.leaseId } : {}),
    },
  );
}

export function leaveWorkflowCollaboration(
  workflowId: string,
  clientId: string,
): Promise<CollaborationSnapshotDTO> {
  return apiSend<CollaborationSnapshotDTO>(
    `${basePath(workflowId)}/presence/${encodeURIComponent(clientId)}`,
    "DELETE",
  );
}

export function leaveWorkflowCollaborationKeepalive(
  workflowId: string,
  clientId: string,
): void {
  void fetch(
    `${basePath(workflowId)}/presence/${encodeURIComponent(clientId)}`,
    {
      method: "DELETE",
      credentials: "include",
      headers: { Accept: "application/json" },
      keepalive: true,
    },
  ).catch(() => undefined);
}

export function workflowCollaborationStreamUrl(
  workflowId: string,
  clientId: string,
): string {
  return withQuery(`${basePath(workflowId)}/stream`, {
    client_id: clientId,
  });
}
